File: src/library/grid/src/just.c
/*
 *  R : A Computer Language for Statistical Data Analysis
 *  Copyright (C) 2001-3 Paul Murrell
 *                2003 The R Core Team
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, a copy is available at
 *  http://www.r-project.org/Licenses/
 */

#include "grid.h"

/* Modify a location for the correct justification */

/* These tranformations assume that x and width are in the same units */
/* FIXME:  I don't think we check anywhere that a horizontal justification
 * is not L_BOTTOM or L_TOP (i.e., meaningless).  Ditto for checking
 * vertical justification.
 */
double justifyX(double x, double width, double hjust) {
    return x - width*hjust;
    /*
     * From when hjust and vjust were enums
     *
    double result = 0;
    switch (hjust) {
    case L_LEFT: 
	result = x;
	break;        
    case L_RIGHT: 
	result = x - width;
	break;
    case L_CENTRE: 
    case L_CENTER:
	result = x - width/2;
	break;
    }
    return result;
    */
}

double justifyY(double y, double height, double vjust) {
    return y - height*vjust;
    /*
     * From when hjust and vjust were enums
     *
    double result = 0;
    switch (vjust) {
    case L_BOTTOM:  
	result = y;
	break;
    case L_TOP:  
	result = y - height;
	break;
    case L_CENTRE:
    case L_CENTER:
	result = y - height/2;
	break;
    }
    return result;
    */
}

/* Convert enum justification into 0..1 justification */
double convertJust(int just) {
    double result = 0;
    switch (just) {
    case L_BOTTOM:
    case L_LEFT:
	result = 0;
	break;
    case L_CENTRE:
    case L_CENTER:
	result = .5;
	break;
    case L_TOP:
    case L_RIGHT:
	result = 1;
	break;
    }
    return result;
}

/* Return the amount of justification required 
 */
void justification(double width, double height, double hjust, double vjust,
		   double *hadj, double *vadj)
{
    *hadj = -width*hjust;
    *vadj = -height*vjust;
    /*
     * From when hjust and vjust were enums
    switch (hjust) {
    case L_LEFT: 
	*hadj = 0;
	break;        
    case L_RIGHT: 
	*hadj = -width;
	break;
    case L_CENTRE: 
    case L_CENTER:
	*hadj = -width/2;
	break;
    }
    switch (vjust) {
    case L_BOTTOM:  
	*vadj = 0;
	break;
    case L_TOP:  
	*vadj = -height;
	break;
    case L_CENTRE:
    case L_CENTER:
	*vadj = -height/2;
	break;
    }
     */
}

File: src/library/grid/R/just.R
#  File src/library/grid/R/just.R
#  Part of the R package, http://www.R-project.org
#
#  Copyright (C) 1995-2013 The R Core Team
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  A copy of the GNU General Public License is available at
#  http://www.r-project.org/Licenses/

# NOTE: the order of the strings in these conversion functions must
# match the order of the enums in ../src/lattice.h
# NOTE: the result of match() is an integer, but subtracting 1 converts
# to real => have to convert back to integer for passing to C code

# If the user specifies two values, the first is horizontal
# justification and the second is vertical

# If the user specifies only one value, use the following
# conversion table to give a second default value
#
# bottom  -->  centre, bottom
# left    -->  left,   centre
# right   -->  right,  centre
# top     -->  centre, top
# centre  -->  centre, centre

valid.charjust <- function(just) {
  if (length(just) == 1) {
    # single value may be any valid just
    just <- as.integer(match(just[1L], c("left", "right", "bottom", "top",
                                        "centre", "center")) - 1)
    if (anyNA(just))
      stop("invalid justification")
  } else if (length(just) > 1) {
    # first value must be one of "left", "right", "centre", or "center"
    just[1L] <- as.integer(match(just[1L], c("left", "right", "bottom", "top",
                                           "centre", "center")) - 1)
    if (!(just[1L] %in% c(0, 1, 4, 5)))
      stop("invalid horizontal justification")
    # second value must be one of "bottom", "top", "centre", or "center"
    just[2L] <- as.integer(match(just[2L], c("left", "right", "bottom", "top",
                                           "centre", "center")) - 1)
    if (!(just[2L] %in% c(2, 3, 4, 5)))
      stop("invalid vertical justification")
    just <- as.integer(just)
  }
  # Extend to length 2 if necessary
  if (length(just) < 2) {
    if (length(just) == 0)
      just <- c(4, 4)
    else
      just <- switch (just[1L] + 1,
                      c(0, 4), # left
                      c(1, 4), # right
                      c(4, 2), # bottom
                      c(4, 3), # top
                      c(4, 4), # centre
                      c(4, 4)) # center
  }
  # Convert to numeric
  just <- c(switch(just[1L] + 1, 0, 1, NA, NA, 0.5, 0.5),
            switch(just[2L] + 1, NA, NA, 0, 1, 0.5, 0.5))
  # Final paranoid check
  if (anyNA(just))
    stop("invalid justification")
  just
}

valid.numjust <- function(just) {
  if (length(just) == 0) {
    c(0.5, 0.5)
  } else {
    if (length(just) < 2) {
      c(just, 0.5)
    } else {
      just
    }
  }
}

valid.just <- function(just) {
  if (is.character(just))
    valid.charjust(just)
  else {
    valid.numjust(as.numeric(just))
  }
}

resolveHJust <- function(just, hjust) {
  if (is.null(hjust) || length(hjust) == 0)
    valid.just(just)[1L]
  else
    hjust
}

resolveVJust <- function(just, vjust) {
  if (is.null(vjust) || length(vjust) == 0)
    valid.just(just)[2L]
  else
    vjust
}
